File: examples/julia/streaming_smoothing.jl
#!/usr/bin/env julia
"""
FastLOESS Streaming Smoothing - Comprehensive Examples

 1. Basic chunked processing
 2. Chunk size comparison
 3. Overlap strategies
 4. Large dataset processing
 5. Outlier handling in streaming mode
 6. File-based streaming simulation
 7. Benchmark (sequential streaming)
 8. Merge strategies
 9. Advanced streaming options
"""

using Printf

# Handle package loading
using Pkg
project_name = Pkg.project().name
if project_name != "FastLOESS"
	script_dir = @__DIR__
	julia_pkg_dir = joinpath(dirname(script_dir), "julia")
	if !haskey(Pkg.project().dependencies, "FastLOESS")
		Pkg.develop(path = julia_pkg_dir)
	end
end

using FastLOESS

make_linear(n) = (collect(Float64, 0:(n-1)), collect(Float64, 0:(n-1)) .* 2 .+ 1)

"""Feed only full-size chunks; finalize() handles remaining data."""
function collect_chunks(model, x, y, chunk_size, overlap)
	step = chunk_size - overlap
	n = length(x)
	result = nothing
	start = 1
	while start + chunk_size - 1 <= n
		res = process_chunk(model, x[start:(start+chunk_size-1)], y[start:(start+chunk_size-1)])
		if result === nothing
			result = res
		else
			append!(result, res)
		end
		start += step
	end
	fin = finalize(model)
	if result === nothing
		result = fin
	else
		append!(result, fin)
	end
	return result
end

# ── Example 1: Basic Chunked Processing ─────────────────────────────────────
function example_1_basic_chunked_processing()
	println("Example 1: Basic Chunked Processing")
	n = 50
	x, y = make_linear(n)
	chunk_size, overlap = 15, 5
	model = StreamingLoess(fraction = 0.5, iterations = 2, chunk_size = chunk_size,
		overlap = overlap, return_residuals = true)
	println("  Dataset: $n pts, chunk=$chunk_size, overlap=$overlap")
	total = 0
	ci = 0
	result = nothing
	start = 1
	while start + chunk_size - 1 <= n
		res = process_chunk(model, x[start:(start+chunk_size-1)], y[start:(start+chunk_size-1)])
		if length(res.x) > 0
			total += length(res.x)
			@printf("  Chunk %d: %d pts (x: %.0f..%.0f)\n", ci, length(res.x), res.x[1], res.x[end])
			result = result === nothing ? res : (append!(result, res); result)
		end
		start += chunk_size - overlap
		ci += 1
	end
	fin = finalize(model)
	if length(fin.x) > 0
		total += length(fin.x)
		@printf("  Finalize: %d remaining pts\n", length(fin.x))
	end
	println("  Total: $total/$n")
	println()
end

# ── Example 2: Chunk Size Comparison ─────────────────────────────────────────
function example_2_chunk_size_comparison()
	println("Example 2: Chunk Size Comparison")
	n = 100
	x, y = make_linear(n)
	for (cs, ov, label) ∈ [(20, 5, "Small"), (50, 10, "Medium"), (80, 15, "Large")]
		model = StreamingLoess(fraction = 0.5, iterations = 1, chunk_size = cs, overlap = ov)
		chunks = 0;
		total = 0;
		start = 1
		while start + cs - 1 <= n
			res = process_chunk(model, x[start:(start+cs-1)], y[start:(start+cs-1)])
			if length(res.x) > 0
				;
				chunks += 1;
				total += length(res.x);
			end
			start += cs - ov
		end
		fin = finalize(model)
		if length(fin.x) > 0
			;
			chunks += 1;
			total += length(fin.x);
		end
		println("  $label (size=$cs, overlap=$ov): chunks=$chunks, total=$total")
	end
	println()
end

# ── Example 3: Overlap Strategies ────────────────────────────────────────────
function example_3_overlap_strategies()
	println("Example 3: Overlap Strategies")
	n = 100
	x, y = make_linear(n)
	cs = 40
	for (overlap, label) ∈ [(0, "No overlap"), (10, "10-pt overlap"), (20, "20-pt overlap")]
		model = StreamingLoess(fraction = 0.5, chunk_size = cs, overlap = overlap)
		total = 0;
		step = cs - overlap;
		start = 1
		while start + cs - 1 <= n
			total += length(process_chunk(model, x[start:(start+cs-1)], y[start:(start+cs-1)]).x)
			start += step
		end
		total += length(finalize(model).x)
		println("  $label: total output=$total")
	end
	println()
end

# ── Example 4: Large Dataset Processing ──────────────────────────────────────
function example_4_large_dataset_processing()
	println("Example 4: Large Dataset Processing")
	n = 10_000
	x = collect(Float64, 0:(n-1))
	y = sin.(x .* 0.01) .+ x .* 0.001
	cs, ov = 500, 50
	model = StreamingLoess(fraction = 0.05, iterations = 2, chunk_size = cs, overlap = ov)
	total = 0;
	step = cs - ov;
	start = 1
	while start + cs - 1 <= n
		total += length(process_chunk(model, x[start:(start+cs-1)], y[start:(start+cs-1)]).x)
		if total > 0 && total % 2000 < step
			println("  Progress: ~$total pts smoothed")
		end
		start += step
	end
	total += length(finalize(model).x)
	println("  Total: $total/$n, memory: constant (chunk=$cs)")
	println()
end

# ── Example 5: Outlier Handling in Streaming Mode ─────────────────────────────
function example_5_outlier_handling()
	println("Example 5: Outlier Handling in Streaming Mode")
	n = 100
	x = collect(Float64, 0:(n-1))
	y = 2 .* x .+ 1 .+ sin.(x .* 0.2) .* 2
	y[[26, 51, 76]] .+= 50  # Outliers (1-indexed)
	for method ∈ ["bisquare", "huber", "talwar"]
		model = StreamingLoess(fraction = 0.5, iterations = 5, robustness_method = method,
			chunk_size = 30, overlap = 10, return_residuals = true)
		large = 0;
		start = 1
		while start + 29 <= n
			res = process_chunk(model, x[start:(start+29)], y[start:(start+29)])
			if res.residuals !== nothing
				large += count(r -> abs(r) > 10, res.residuals)
			end
			start += 20
		end
		fin = finalize(model)
		if fin.residuals !== nothing
			large += count(r -> abs(r) > 10, fin.residuals)
		end
		println("  $method: pts with |residual|>10: $large")
	end
	println()
end

# ── Example 6: File-Based Streaming Simulation ───────────────────────────────
function example_6_file_simulation()
	println("Example 6: File-Based Streaming Simulation")
	println("  Simulating: input.csv -> Smooth -> output.csv")
	total_lines, cs, ov = 200, 50, 10
	model = StreamingLoess(fraction = 0.5, iterations = 2, chunk_size = cs,
		overlap = ov, return_residuals = true)
	out_count = 0;
	ci = 0;
	start_line = 0
	while start_line < total_lines
		end_line = min(start_line + cs, total_lines)
		xc = collect(Float64, start_line:(end_line-1))
		yc = 2 .* xc .+ 1 .+ sin.(xc .* 0.1) .* 3
		println("  Reading chunk $ci (lines $start_line..$(end_line-1))")
		res = process_chunk(model, xc, yc)
		if length(res.x) > 0
			out_count += length(res.x)
			println("    -> Writing $(length(res.x)) smoothed pts (total: $out_count)")
		end
		start_line += cs - ov
		ci += 1
	end
	fin = finalize(model)
	if length(fin.x) > 0
		out_count += length(fin.x)
		println("  Finalizing: $(length(fin.x)) remaining pts")
	end
	println("  Input: $total_lines, Output: $out_count")
	println()
end

# ── Example 7: Benchmark (Sequential Streaming) ───────────────────────────────
function example_7_benchmark()
	println("Example 7: Benchmark (Sequential Streaming)")
	n, cs, ov = 1000, 100, 10
	model = StreamingLoess(fraction = 0.5, iterations = 3, chunk_size = cs, overlap = ov)
	t0 = time()
	total = 0;
	start = 1
	while start + cs - 1 <= n
		xc = collect(Float64, (start-1):(start+cs-2))
		yc = sin.(xc .* 0.1) .+ cos.(xc .* 0.01)
		total += length(process_chunk(model, xc, yc).x)
		start += cs - ov
	end
	total += length(finalize(model).x)
	ms = (time() - t0) * 1000
	@printf("  %d pts in %.2fms (chunk=%d, overlap=%d)\n", total, ms, cs, ov)
	println()
end

# ── Example 8: Merge Strategies ──────────────────────────────────────────────
function example_8_merge_strategies()
	println("Example 8: Merge Strategies")
	n = 50
	x, y = make_linear(n)
	for strategy ∈ ["average", "weighted_average", "take_first", "take_last"]
		model = StreamingLoess(fraction = 0.5, iterations = 2, chunk_size = 20,
			overlap = 5, merge_strategy = strategy)
		total = 0;
		start = 1
		while start + 19 <= n
			total += length(process_chunk(model, x[start:(start+19)], y[start:(start+19)]).x)
			start += 15
		end
		total += length(finalize(model).x)
		println("  $strategy: total=$total")
	end
	println()
end

# ── Example 9: Advanced Streaming Options ─────────────────────────────────────
function example_9_advanced_options()
	println("Example 9: Advanced Streaming Options")
	n = 50
	x, y = make_linear(n)
	model = StreamingLoess(
		fraction = 0.5, iterations = 2,
		degree = "quadratic",
		scaling_method = "mar",
		boundary_policy = "reflect",
		zero_weight_fallback = "return_original",
		distance_metric = "manhattan",
		surface_mode = "direct",
		return_se = true,
		return_diagnostics = true,
		return_robustness_weights = true,
		auto_converge = 1e-3,
		chunk_size = 20, overlap = 5,
	)
	total = 0;
	start = 1
	while start + 19 <= n
		total += length(process_chunk(model, x[start:(start+19)], y[start:(start+19)]).x)
		start += 15
	end
	fin = finalize(model)
	total += length(fin.x)
	println("  total pts: $total")
	if fin.standard_errors !== nothing && !isempty(fin.standard_errors)
		@printf("  standard_errors[1]: %.4f\n", fin.standard_errors[1])
	end
	if fin.diagnostics !== nothing
		@printf("  diagnostics.rmse: %.3f\n", fin.diagnostics.rmse)
		@printf("  diagnostics.r_squared: %.3f\n", fin.diagnostics.r_squared)
		!isnan(fin.diagnostics.aic) && @printf("  diagnostics.aic: %.3f\n", fin.diagnostics.aic)
	end
	if fin.robustness_weights !== nothing && !isempty(fin.robustness_weights)
		@printf("  robustness_weights[1]: %.4f\n", fin.robustness_weights[1])
	end
	println()
end

# ── Main ──────────────────────────────────────────────────────────────────────
function main()
	println("=" ^ 60)
	println("FastLOESS Streaming Smoothing - Comprehensive Examples")
	println("=" ^ 60)
	println()

	example_1_basic_chunked_processing()
	example_2_chunk_size_comparison()
	example_3_overlap_strategies()
	example_4_large_dataset_processing()
	example_5_outlier_handling()
	example_6_file_simulation()
	example_7_benchmark()
	example_8_merge_strategies()
	example_9_advanced_options()

	println("=== Streaming Smoothing Examples Complete ===")
end

main()

File: examples/nodejs/streaming_smoothing.js
const fastloess = require('../../bindings/nodejs');

/**
 * fastloess Streaming Smoothing - Comprehensive Examples
 *
 * 9 examples covering the full StreamingLoess API:
 *  1. Basic chunked processing
 *  2. Chunk size comparison
 *  3. Overlap strategies
 *  4. Large dataset processing
 *  5. Outlier handling in streaming mode
 *  6. File-based streaming simulation
 *  7. Benchmark (sequential streaming)
 *  8. Merge strategies (average, weighted_average, take_first, take_last)
 *  9. Advanced streaming options
 */

function makeLinear(n) {
    const x = new Float64Array(n);
    const y = new Float64Array(n);
    for (let i = 0; i < n; i++) { x[i] = i; y[i] = 2 * i + 1; }
    return { x, y };
}

// ── Example 1: Basic Chunked Processing ─────────────────────────────────────
function example_1_basic_chunked_processing() {
    console.log("Example 1: Basic Chunked Processing");

    const n = 50;
    const { x, y } = makeLinear(n);
    const chunk_size = 15;
    const overlap = 5;

    const streamer = new fastloess.StreamingLoess(
        { fraction: 0.5, iterations: 2, return_residuals: true },
        { chunk_size, overlap }
    );

    console.log(`  Dataset: ${n} points, chunk=${chunk_size}, overlap=${overlap}`);

    let totalProcessed = 0;
    let chunkIdx = 0;
    for (let start = 0; start < n; start += chunk_size - overlap) {
        const end = Math.min(start + chunk_size, n);
        const res = streamer.process_chunk(x.subarray(start, end), y.subarray(start, end));
        if (res.x.length > 0) {
            totalProcessed += res.x.length;
            console.log(`  Chunk ${chunkIdx}: ${res.x.length} pts (x: ${res.x[0].toFixed(0)}..${res.x[res.x.length - 1].toFixed(0)})`);
        }
        chunkIdx++;
    }
    const fin = streamer.finalize();
    if (fin.x.length > 0) {
        totalProcessed += fin.x.length;
        console.log(`  Finalize: ${fin.x.length} remaining pts`);
    }
    console.log(`  Total: ${totalProcessed}/${n}`);
    console.log();
}

// ── Example 2: Chunk Size Comparison ────────────────────────────────────────
function example_2_chunk_size_comparison() {
    console.log("Example 2: Chunk Size Comparison");

    const n = 100;
    const { x, y } = makeLinear(n);

    for (const [chunk_size, overlap, label] of [[20, 5, "Small"], [50, 10, "Medium"], [80, 15, "Large"]]) {
        const streamer = new fastloess.StreamingLoess(
            { fraction: 0.5, iterations: 1 },
            { chunk_size, overlap }
        );
        let chunks = 0, total = 0;
        for (let start = 0; start < n; start += chunk_size - overlap) {
            const end = Math.min(start + chunk_size, n);
            const res = streamer.process_chunk(x.subarray(start, end), y.subarray(start, end));
            if (res.x.length > 0) { chunks++; total += res.x.length; }
        }
        const fin = streamer.finalize();
        if (fin.x.length > 0) { chunks++; total += fin.x.length; }
        console.log(`  ${label} (size=${chunk_size}, overlap=${overlap}): chunks=${chunks}, total=${total}`);
    }
    console.log();
}

// ── Example 3: Overlap Strategies ────────────────────────────────────────────
function example_3_overlap_strategies() {
    console.log("Example 3: Overlap Strategies");

    const n = 100;
    const { x, y } = makeLinear(n);

    for (const [overlap, label] of [[0, "No overlap"], [10, "10-pt overlap"], [20, "20-pt overlap"]]) {
        const chunk_size = 40;
        const streamer = new fastloess.StreamingLoess(
            { fraction: 0.5 },
            { chunk_size, overlap }
        );
        let total = 0;
        const step = chunk_size - overlap;
        // Feed only full-size chunks; finalize() handles remaining data
        for (let start = 0; start + chunk_size <= n; start += step) {
            const res = streamer.process_chunk(x.subarray(start, start + chunk_size), y.subarray(start, start + chunk_size));
            total += res.x.length;
        }
        total += streamer.finalize().x.length;
        console.log(`  ${label} (overlap=${overlap}): total points output=${total}`);
    }
    console.log();
}

// ── Example 4: Large Dataset Processing ──────────────────────────────────────
function example_4_large_dataset_processing() {
    console.log("Example 4: Large Dataset Processing");

    const n = 10000;
    const x = new Float64Array(n);
    const y = new Float64Array(n);
    for (let i = 0; i < n; i++) { x[i] = i; y[i] = Math.sin(i * 0.01) + i * 0.001; }

    const chunk_size = 500;
    const overlap = 50;

    const streamer = new fastloess.StreamingLoess(
        { fraction: 0.05, iterations: 2 },
        { chunk_size, overlap }
    );

    let total = 0;
    const step = chunk_size - overlap;
    for (let start = 0; start < n; start += step) {
        const end = Math.min(start + chunk_size, n);
        const res = streamer.process_chunk(x.subarray(start, end), y.subarray(start, end));
        total += res.x.length;
        if (total > 0 && total % 2000 < step) {
            console.log(`  Progress: ~${total} points smoothed`);
        }
    }
    total += streamer.finalize().x.length;
    console.log(`  Total processed: ${total}/${n}`);
    console.log(`  Memory efficiency: constant (chunk=${chunk_size})`);
    console.log();
}

// ── Example 5: Outlier Handling in Streaming Mode ─────────────────────────────
function example_5_outlier_handling() {
    console.log("Example 5: Outlier Handling in Streaming Mode");

    const n = 100;
    const x = new Float64Array(n);
    const y = new Float64Array(n);
    for (let i = 0; i < n; i++) {
        x[i] = i;
        y[i] = 2 * i + 1 + Math.sin(i * 0.2) * 2;
        if (i === 25 || i === 50 || i === 75) y[i] += 50; // Outliers
    }

    for (const method of ["bisquare", "huber", "talwar"]) {
        const streamer = new fastloess.StreamingLoess(
            { fraction: 0.5, iterations: 5, robustness_method: method, return_residuals: true },
            { chunk_size: 30, overlap: 10 }
        );
        let largeResiduals = 0;
        for (let start = 0; start < n; start += 20) {
            const end = Math.min(start + 30, n);
            const res = streamer.process_chunk(x.subarray(start, end), y.subarray(start, end));
            if (res.residuals) {
                for (const r of res.residuals) { if (Math.abs(r) > 10) largeResiduals++; }
            }
        }
        const fin = streamer.finalize();
        if (fin.residuals) {
            for (const r of fin.residuals) { if (Math.abs(r) > 10) largeResiduals++; }
        }
        console.log(`  ${method}: points with |residual|>10: ${largeResiduals}`);
    }
    console.log();
}

// ── Example 6: File-Based Streaming Simulation ───────────────────────────────
function example_6_file_simulation() {
    console.log("Example 6: File-Based Streaming Simulation");
    console.log("  Simulating: Read from input.csv -> Smooth -> Write to output.csv");

    const totalLines = 200;
    const chunk_size = 50;
    const overlap = 10;

    const streamer = new fastloess.StreamingLoess(
        { fraction: 0.5, iterations: 2, return_residuals: true },
        { chunk_size, overlap }
    );

    let outputLines = 0;
    for (let ci = 0; ci < Math.ceil(totalLines / (chunk_size - overlap)); ci++) {
        const start = ci * (chunk_size - overlap);
        const end = Math.min(start + chunk_size, totalLines);

        // Simulate reading a chunk from a file
        const xChunk = new Float64Array(end - start);
        const yChunk = new Float64Array(end - start);
        for (let j = 0; j < end - start; j++) {
            xChunk[j] = start + j;
            yChunk[j] = 2 * xChunk[j] + 1 + Math.sin(xChunk[j] * 0.1) * 3;
        }

        console.log(`  Reading chunk ${ci} (lines ${start}..${end - 1})`);
        const res = streamer.process_chunk(xChunk, yChunk);
        if (res.x.length > 0) {
            outputLines += res.x.length;
            console.log(`    -> Writing ${res.x.length} smoothed pts (total: ${outputLines})`);
        }
    }

    const fin = streamer.finalize();
    if (fin.x.length > 0) {
        outputLines += fin.x.length;
        console.log(`  Finalizing: Writing ${fin.x.length} remaining pts`);
    }
    console.log(`  Input: ${totalLines}, Output: ${outputLines}`);
    console.log();
}

// ── Example 7: Benchmark (Sequential Streaming) ───────────────────────────────
function example_7_benchmark() {
    console.log("Example 7: Benchmark (Sequential Streaming)");

    const n = 1000;
    const chunk_size = 100;
    const overlap = 10;

    const streamer = new fastloess.StreamingLoess(
        { fraction: 0.5, iterations: 3 },
        { chunk_size, overlap }
    );

    const t0 = process.hrtime.bigint();
    let total = 0;
    const step = chunk_size - overlap;
    for (let start = 0; start < n; start += step) {
        const end = Math.min(start + chunk_size, n);
        const xc = new Float64Array(end - start);
        const yc = new Float64Array(end - start);
        for (let j = 0; j < end - start; j++) {
            xc[j] = start + j;
            yc[j] = Math.sin(xc[j] * 0.1) + Math.cos(xc[j] * 0.01);
        }
        total += streamer.process_chunk(xc, yc).x.length;
    }
    total += streamer.finalize().x.length;
    const ms = Number(process.hrtime.bigint() - t0) / 1e6;

    console.log(`  ${total} points in ${ms.toFixed(2)}ms`);
    console.log(`  chunk=${chunk_size}, overlap=${overlap}`);
    console.log();
}

// ── Example 8: Merge Strategies ──────────────────────────────────────────────
function example_8_merge_strategies() {
    console.log("Example 8: Merge Strategies");

    const n = 50;
    const { x, y } = makeLinear(n);

    for (const strategy of ["average", "weighted_average", "take_first", "take_last"]) {
        const streamer = new fastloess.StreamingLoess(
            { fraction: 0.5, iterations: 2 },
            { chunk_size: 20, overlap: 5, merge_strategy: strategy }
        );
        let total = 0;
        for (let start = 0; start < n; start += 15) {
            const end = Math.min(start + 20, n);
            total += streamer.process_chunk(x.subarray(start, end), y.subarray(start, end)).x.length;
        }
        total += streamer.finalize().x.length;
        console.log(`  ${strategy}: total=${total}`);
    }
    console.log();
}

// ── Example 9: Advanced Streaming Options ─────────────────────────────────────
function example_9_advanced_options() {
    console.log("Example 9: Advanced Streaming Options");

    const n = 50;
    const { x, y } = makeLinear(n);

    const streamer = new fastloess.StreamingLoess(
        {
            fraction: 0.5,
            iterations: 2,
            degree: "quadratic",
            scaling_method: "mar",
            boundary_policy: "reflect",
            zero_weight_fallback: "return_original",
            distance_metric: "manhattan",
            surface_mode: "direct",
            return_se: true,
            return_diagnostics: true,
            return_robustness_weights: true,
            auto_converge: 1e-3,
        },
        { chunk_size: 20, overlap: 5 }
    );

    let total = 0;
    for (let start = 0; start < n; start += 15) {
        const end = Math.min(start + 20, n);
        total += streamer.process_chunk(x.subarray(start, end), y.subarray(start, end)).x.length;
    }
    const fin = streamer.finalize();
    total += fin.x.length;

    console.log(`  total points: ${total}`);
    if (fin.standard_errors && fin.standard_errors.length > 0) {
        console.log(`  standard_errors[0]: ${fin.standard_errors[0].toFixed(4)}`);
    }
    if (fin.diagnostics) {
        console.log(`  diagnostics.rmse: ${fin.diagnostics.rmse.toFixed(3)}`);
        console.log(`  diagnostics.r_squared: ${fin.diagnostics.r_squared.toFixed(3)}`);
        if (fin.diagnostics.aic != null) console.log(`  diagnostics.aic: ${fin.diagnostics.aic.toFixed(3)}`);
    }
    if (fin.robustness_weights && fin.robustness_weights.length > 0) {
        console.log(`  robustness_weights[0]: ${fin.robustness_weights[0].toFixed(4)}`);
    }
    console.log();
}

// ── Main ──────────────────────────────────────────────────────────────────────
function main() {
    console.log("=".repeat(60));
    console.log("fastloess Streaming Smoothing - Comprehensive Examples");
    console.log("=".repeat(60));
    console.log();

    example_1_basic_chunked_processing();
    example_2_chunk_size_comparison();
    example_3_overlap_strategies();
    example_4_large_dataset_processing();
    example_5_outlier_handling();
    example_6_file_simulation();
    example_7_benchmark();
    example_8_merge_strategies();
    example_9_advanced_options();

    console.log("=== Streaming Smoothing Examples Complete ===");
}

main();

